src/coreclr/jit/optimizer.cpp

```diff
@@ -9316,6 +9318,227 @@ bool OptBoolsDsc::optOptimizeBoolsCondBlock()
+//
+//      ------------ BB01 -> BB03 (cond), succs={BB03,BB04}
+//      *  JTRUE
+//      \--*  AND
+//         +--*  LT a,b
+//         \--*  NE c,d
+//
+//      ------------ BB03, preds={BB01} succs={BB05}
+//      *  ASG x,y
+//
+//      ------------ BB04, preds={BB01} succs={BB05}
+//      *  ASG x,z
+//
+//
+//      Multiple conditions can be chained together. This is due to the optimization reverse
+//      iterating through the blocks. For example:
+//          If ( a > b || c == d || e < f ) { x = y; }
+//      The first pass will combine "c == d" and "e < f" into a chain. The second pass will then
+//      combine the "a > b" with the earlier chain, giving:
+//
+//      ------------ BB01 -> BB03 (cond), succs={BB03,BB04}
+//      *  JTRUE
+//      \--*  AND
+//         +--*  AND
+//            +--*  NE c,d
+//            +--*  GE e,f
+//         \--*  LT a,b
+//
+//      ------------ BB03, preds={BB01} succs={BB04}
+//      *  ASG x,y
+//
+//
+//      Conditions connected by && are not yet checked for. For example:
+//          If ( a > b && c == d ) { x = y; }
+//
+bool OptBoolsDsc::optOptimizeCompareChainCondBlock()
+{
+    assert(m_b1 != nullptr && m_b2 != nullptr && m_b3 == nullptr);
+    m_t3 = nullptr;
+
+    if (!(m_b1->bbNext == m_b2 && m_b1->bbJumpDest == m_b2->bbNext && m_b1->bbJumpDest->bbNext == m_b2->bbJumpDest))
+    {
+        return false;
+    }
+
+    Statement* const s1 = optOptimizeBoolsChkBlkCond();
+    if (s1 == nullptr)
+    {
+        return false;
+    }
+    Statement* s2 = m_b2->firstStmt();
+
+    assert(m_testInfo1.testTree->OperIs(GT_JTRUE));
+    GenTree* cond1 = m_testInfo1.testTree->gtGetOp1();
+    assert(m_testInfo2.testTree->OperIs(GT_JTRUE));
+    GenTree* cond2 = m_testInfo2.testTree->gtGetOp1();
+
+    // Ensure both conditions are suitable.
+    if (!cond1->OperIsCmpCompare())
+    {
+        return false;
+    }
+    if (!(cond2->OperIsCmpCompare() || cond2->OperIs(GT_AND)))
+    {
+        return false;
+    }
+
+    // Ensure there are no additional side effects.
+    if ((cond1->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0 ||
+        (cond2->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+    {
+        return false;
+    }
+
+    // Integer compares only for now (until support for Arm64 fccmp instruction is added)
+    if (varTypeIsFloating(cond1->gtGetOp1()->TypeGet()) || varTypeIsFloating(cond2->gtGetOp1()->TypeGet()))
+    {
+        return false;
+    }
+
+    // Avoid cases where the compare will be optimized better later:
+    // * cmp(and(x, y), 0) will be turned into a TEST_ opcode.
+    // * Compares against zero will be optimized with cbz.
+    // Make sure to avoid matching previous optimize bool cases.
+    bool op1IsCondChain = false;
+    bool op2IsCondChain = false;
+    if (ConditionIsTest(cond1, &op1IsCondChain) || ConditionIsTest(cond2, &op2IsCondChain))
+    {
+        return false;
+    }
+
```

Comment:
Does the code need some kind of profitability (as opposed to legality) check here? Is it possible for the second compare to be an "expensive" operation that might be avoided if the first condition in a short-circuit evaluation is true and thus avoids it?

Reply:
Even if the conditions are all relatively cheap, is it possible to construct a very long chain where the first (short-circuit) condition is 99% likely to be true so the subsequent conditions rarely need to be executed? Should the code here consider block/edge weight (PGO) data?

Reply:
Yes, it's possible to construct an arbitrary long chain here. In reality, I wasn't seeing anything too long in the diffs.

Do we have any data to know how likely a condition is going to be? Even if the first condition is very likely, then chaining a few conditions is still better than having all the branching. LLVM limits to chains of three, which is probably a good starting point.

Reply:
Block weights should give this data when we have PGO information (`fgPgoHaveWeights`). However, I'm not sure how much we can rely on the PGO data at this point in the JIT -- @AndyAyersMS is currently working on making sure that the PGO data is handled correctly in each individual phase.

I think your suggestion on a limit makes sense, with some added heuristics like if-conversion has. This is similar to if-conversion in that it turns conditionally executed code into unconditionally executed code, and we picked the heuristic there so that we increase the total cost by up to the cost of `ADD(LCL_VAR, LCL_VAR)` (which was `7` IIRC). So we could e.g. say that a tree like `EQ(ADD(LCL_VAR, LCL_VAR), LCL_VAR)` is ok to unconditionally execute, but no more complex than that.

I also think stress modes similar to if-conversion here are going to be beneficial.

I'm looking forward to seeing the diffs of this change. I think merging this should be simpler now assuming no backend changes are necessary anymore.

Reply:
I'll look at adding something very similar to if conversion for the limiting.
I'll be doing the merge up to latest head on Thursday. Hoping it should be fairly simple now.

Reply:
I've added some costing, but I'm not convinced it's ideal. If one of the conditions is already multiple conditions, then the cost is quite large. Looking at the diffs, the costing isn't preventing much, which is good.